fhir-r4/src/main/java/org/dhis2/fhir/adapter/fhir/transform/dhis/impl/aggregate/r4/R4FhirMeasureReportToDataValueSetTransformer.java

```diff
@@ -0,0 +1,208 @@
+        throws TransformerException
+    {
+        final Map<String, Object> variables = new HashMap<>( scriptVariables );
+        final MeasureReport mr = (MeasureReport) input;
+
+        if ( ruleInfo.getRule().getDataSetIdLookupScript() == null )
+        {
+            logger.info( "Rule does not define a data value set lookup script and data value set does not yet include one." );
+            return null;
+        }
+        else
+        {
+            Optional<String> dataSetId = getDataSetId( context, ruleInfo, ruleInfo.getRule().getDataSetIdLookupScript(), variables );
+            dataSetId.ifPresent( dataValueSet::setDataSetId );
+        }
+
+        if ( dataValueSet.getDataSetId() == null )
+        {
+            throw new TransformerDataException( "Data Value Set ID cannot be decided." );
+        }
+
+        try
+        {
+            if ( Objects.isNull( mr.getPeriod() ) || Objects.isNull( mr.getPeriod().getStart() ) || Objects.isNull( mr.getPeriod().getEnd() ) )
+            {
+                throw new TransformerDataException( "Period start and/or end is not provided. Period cannot be decided." );
+            }
+
+            dataValueSet.setPeriod( PeriodUtils.getDHIS2PeriodString( mr.getPeriod().getStart(), mr.getPeriod().getEnd() ) );
+        }
+        catch ( IllegalArgumentException e )
+        {
+            throw new TransformerDataException( e.getMessage() );
+        }
+
+        ZonedDateTime lastUpdated = ZonedDateTime.ofInstant( mr.getMeta().getLastUpdated().toInstant(), zoneId );
+        transformDataValues( mr, dataValueSet, lastUpdated );
+        //        String version = mr.getMeta().getVersionId();
+
+        dataValueSet.setId( createArtificialDataValueSetId( mr ) );
+
+        return new FhirToDhisTransformOutcome<>( ruleInfo.getRule(), dataValueSet, dataValueSet.isNewResource() );
+    }
+
+    private String createArtificialDataValueSetId( MeasureReport measureReport )
+    {
+        try
+        {
+            String measureUuid = measureReport.getMeasure().substring( 8 );
+            String locationId = measureReport.getReporter().getReference().substring( 9 );
+            String period = PeriodUtils.getDHIS2PeriodString( measureReport.getPeriod().getStart(), measureReport.getPeriod().getEnd() );
+
+            //DataValueSet doesn't use IDs. Therefore, creating the artificial one only for the Adapter purposes. It consists of:
+            //measure UUID (equivalent of DHIS2 DataSetId), location UUID (equivalent of DHIS2 OrgUnitUID) and period in DHIS2 format
+            return measureUuid + ":" + locationId + ":" + period;
+        }
+        catch ( IllegalArgumentException e )
+        {
+            throw new TransformerDataException( e.getMessage() );
+        }
+    }
+
+    private void transformDataValues( MeasureReport measureReport, DataValueSet dataValueSet, ZonedDateTime lastUpdated )
+    {
+        for ( int i = 1; i < measureReport.getGroup().size(); i++ )
```

Comment:
Does it make any sense to use this kind of loop instead of the loop over an iterable? And why to skip the first group item???